Regex in Gateway API (#4764)
kubernetes-sigs/gateway-api: geps/gep-4359/index.md

# GEP-4359: Gateway API Regex

* Issue: [#4359](https://github.com/kubernetes-sigs/gateway-api/issues/4359)
* Status: Provisional

## TLDR

Regular expressions provide a powerful and concise method for traffic routing and manipulation, but present portability challenges due to the variety of regex engines.

> The goal of this proposal is to define a common regex syntax and semantics that all implementations should support.

Reviewer: ```suggestion
The goal of this proposal is to define a common minimum regex syntax and semantics that all implementations should support.
```
It sounds like we are actually defining a minimum set of features that are required, and we will test. But we are explicitly avoiding defining a maximum?

Reviewer 2: I think we should avoid defining any minimum and maximum set of features. I believe we should mirror the current behavior of regex matchers where it's implementation specific. That would keep the API transparent and consistent across multiple features.

Reviewer 2: ah, I see, this proposal aims to change exactly this behaviour.

Null and control characters should not show up in traffic routing.

Some examples of valid input strings are

* `foo` (`0x66 0x6F 0x6F`)
* ` foo bar` (`0x20 0x66 0x6F 0x6F 0x20 0x62 0x61 0x72`)
* `foo\tbar` (`0x66 0x6F 0x6F 0x09 0x62 0x61 0x72`)

Some examples of invalid input strings are (implementations need not reject these, but they are not explicitly supported by the spec):

* `foo\nbar` (`0x66 0x6F 0x6F 0x0A 0x62 0x61 0x72`)
* `ｆ๏๏ ß∆я` (`0xef 0xbd 0x86 0xe0 0xb9 0x8f 0xe0 0xb9 0x8f 0xe2 0x80 0x82 0xc3 0x9f 0xe2 0x88 0x86 0xd1 0x8f`)

## Gateway API Regex Definition

We use IEEE POSIX ERE (as defined in Chapter 9 of the [The Open Group Base Specifications Issue 8]), as the base of Gateway API Regex, with a few exceptions:

* We do not define whether matches are leftmost-longest, leftmost-first, or something else.
* Unescaped square brackets in character classes are undefined (`[[]]`, `[]]`, `[[]`, `[[:alpha:]`, `[[=a=]]`, `[[.ch.]]`).
* Backslashes (`\`) retain their special meaning in character classes and can escape special characters, even if those special characters lose their special meaning in character classes. For example, `[\-]` matches `-`, and `[\]]` matches `]`.

> * Consecutive quantifiers are undefined (`a**`, `a*?`, `a{1,2}?`).

Reviewer 2: it might be a brealking change for existing regex-based matchers. For example, `a{1,2}?` is a valid expression in RE2, thus making it undefined, we immediately break existing Envoy users.

Author: Maybe I can make this more explicit, but when I say undefined, I don't mean that the control plane shouldn't accept the regex, its more that there are no portability guarantees. I don't think `a{1,2}?` is a particularly portable. the `?` can either modify the previous quantifier and make it lazy, or it can take its usual meaning of "0 or 1"

Reviewer 2: right, maybe we can use term "implementation specific"? I think with this phrasing it would be even more appealing: we standardize a common subset of  regex syntax, while everything else is implementation specific.; as compared to "everything is implementation specific".

Reviewer 3: For "undefined" vs "implementation specific", I have a slight preference for "undefined" -- I'd rather encourage folks to stick to the standard.

However -- if I'm reading regex101.com correctly (that is an _amazing_ tool), then `a{1,2}` seems to be supported by all the engines we care about, so perhaps we can just include it? @Stevenjin8, was there an engine you found where it's not supported? 🤔

Author: its `a{1,2}?` (emphasis on the trailing question mark). Most engines agree that this means "lazily match 1 or two of the character `a`", but this is not the case with POSIX ERE and I don't think this is common (its not listed under the "Common tokens" tab in regex101. So I'd rather leave this behavior undefined. We can add it later, but my goal with this PR is to stick as closely to POSIX ERE while still being broadly compatible with modern engines.

Reviewer 3: 🤦‍♂️ Oh, that's just embarrassing. Yes, completely agree that `a{1,2}` should be part of the standard, but that `a{1,2}?` should not. Sorry for the confusion!

* Any locale-specific behavior will assume the [C/POSIX locale](https://pubs.opengroup.org/onlinepubs/7908799/xbd/locale.html) (e.g. character ordering).

IEEE POSIX ERE is a good common denominator because
* The set of supported features is small (e.g. no backreferences)
* Broadly compatible across regex engines (RE2, rust-lang/regex, PCRE), especially because we don't have to worry about unicode, line breaks, and control sequences.

> Some notable omissions from Gateway API Regex are:

Reviewer 2: same as above - many of these operators are currently supported by this API; omitting them now would lead to a breaking change.

Author: This API doesn't exist. I don't see how it could be a breaking change. Regardless, this is more of a "follow these rules to guarantee portability", not a limit on the syntax gwapi implementations can support.

Reviewer 2: I meant existing regex matchers - do we plan to apply these limitations there as well? My understanding of this GEP is to treat certain syntax (that currently exists and supported, such as `a{1,2}?`) as an undefined behaviour.

If this is wrong impression, then we might need to rephrase it, or make it more explicit that some implementations will support a full syntax.

Author: yeah, I'll just clear it up in the GEP

* `-`
* ` `

The pattern `[^a-zA-Z]` would match

* `-`
* ` `
* `9`

The following strings would not match

* `a`
* `Z`
* `k`

For the pattern `^[^^]$`, the only non-matching strings are

* ``
* `^`
* Any string with two or more characters

For the patterns `^[-a]$` and `^[a-]$`, the only matching strings are 

* `-`
* `a`

For the equivalent patterns `^[-.*(){}|^$\\]$` and `^[\-\.\*\(\)\{\}\|\^\$\\]$`, the only matching strings are

* `-`
* `.`
* `*`
* `(`
* `)`
* `{`
* `}`
* `|`
* `^`
* `$`
* `\`

For the pattern `^[\-Z]$`, the only matching strings are

* `-`
* `Z`

## Replacement

Replacement is usually straightforward, but has some semantic ambiguities.
Given a pattern, a replacement, and an input, we allow higher level APIs to define semantics such as
* whether the first or all matches should be replaced.
* if replacing all matches, how to deal with overlapping matches.

That said, whether the matches are leftmost-first or leftmost-longest is always undefined.

### Referencing Capturing Groups

In the case that referencing capturing groups is supported, APIs MUST use the `\1`, `\2`, etc. syntax in the replacement string to reference capturing groups.

> Named capturing groups are not explicitly supported, and the syntax for referencing them is undefined.

Reviewer 2: why?

Author: 1. I don't think the syntax is particularly standardized
2. Named capturing groups are purely ergonomic
3. I think its better that we start off with a limit set of supported features as this is supposed to be a common denominator.

Reviewer 2: I don't have strong arguments at the moment, I just look at the possible evolution of the regex rewrite section where we can use some capture groups from request header and use them to build a new path. Agree, this is pretty niche feature, but numbered capture groups would make this scenario impossible: consider we have two patterns for each header and path, each pattern will have \1 \2 \3 capture groups - we could not distinguish the numbered groups in substitute.

I know that here we do not reject named groups, but with the possible regex rewrite evolution above, numbered capture groups might be rejected by the vendor's API.

I just don't feel like we want to standardize numbered capture groups. It might create more problems in future than it solves.

Reviewer 4: I was actually wondering if named captures would always be accepted but not the numbered ones. I was thinking that something like this would make sense:

```yaml 
- regex: {protocol}://{fqdn}({port})?/{path1}/{path2}/{path3}?{arg1}&{arg2}
  replace: https://{fqdn}/{path2}/{path1}?{arg2}&something=else
```

mostly thinking that regexes can be "hard to define" initially, and trying to understand if the majority of user stories are just related to something like "I want to remove a path", "I want to add an extra path", "I want to inject an argument"

Reviewer 3: I'm with @Stevenjin8 on this one: I'd prefer something that can be directly passed to the engine without preprocessing, and that implies numbered groups but not named groups.

I'm not completely opposed to defining our own named-group syntax (as in @rikatz's example), but I'm not sure how valuable it'll be since it is _by definition_ going to be a breaking change that will affect users.

Reviewer 3: ...I should've phrased that last point as "it feels likely that requiring named groups will force more people to change what they do than allowing only numbered groups". As @Stevenjin8 points out, when there's no API, you can't really have a "breaking change".

Reviewer 2: I agree, I think I misunderstood this is a required limitation rather than required minimum. I think named groups might still remain implementation-specific, while numbered groups - minimum requirement.

Reviewer: Remember that every time you add an implementation-specific thing, you build a migration footgun for your users.

Some examples of valid input strings are

* `foo` (`0x66 0x6F 0x6F`)
* ` foo bar` (`0x20 0x66 0x6F 0x6F 0x20 0x62 0x61 0x72`)
* `foo\tbar` (`0x66 0x6F 0x6F 0x09 0x62 0x61 0x72`)

Some examples of invalid input strings are (implementations need not reject these, but they are not explicitly supported by the spec):

* `foo\nbar` (`0x66 0x6F 0x6F 0x0A 0x62 0x61 0x72`)
* `ｆ๏๏ ß∆я` (`0xef 0xbd 0x86 0xe0 0xb9 0x8f 0xe0 0xb9 0x8f 0xe2 0x80 0x82 0xc3 0x9f 0xe2 0x88 0x86 0xd1 0x8f`)

## Gateway API Regex Definition

We use IEEE POSIX ERE (as defined in Chapter 9 of the [The Open Group Base Specifications Issue 8]), as the base of Gateway API Regex, with a few exceptions:

* We do not define whether matches are leftmost-longest, leftmost-first, or something else.
* Unescaped square brackets in character classes are undefined (`[[]]`, `[]]`, `[[]`, `[[:alpha:]`, `[[=a=]]`, `[[.ch.]]`).
* Backslashes (`\`) retain their special meaning in character classes and can escape special characters, even if those special characters lose their special meaning in character classes. For example, `[\-]` matches `-`, and `[\]]` matches `]`.
* Consecutive quantifiers are undefined (`a**`, `a*?`, `a{1,2}?`).

> * Any locale-specific behavior will assume the [C/POSIX locale](https://pubs.opengroup.org/onlinepubs/7908799/xbd/locale.html) (e.g. character ordering).

Reviewer 4: any definition for case sensitive/insensitive behavior?

Reviewer 4: ah just saw below

# GEP-4359: Gateway API Regex

* Issue: [#4359](https://github.com/kubernetes-sigs/gateway-api/issues/4359)
* Status: Provisional

## TLDR

Regular expressions provide a powerful and concise method for traffic routing and manipulation, but present portability challenges due to the variety of regex engines.
The goal of this proposal is to define a common regex syntax and semantics that all implementations should support to ensure portability, but mostly testability.
We use POSIX ERE (as defined in Chapter 9 of the [The Open Group Base Specifications Issue 8]) as the common denominator, with some feature exclusions.

## Goals

* Define a common regex syntax and semantics (Gateway API Regex) for Gateway API that all implementations should support.
* Define the set of inputs for our regex expressions.
* Define methods of referencing capturing groups.

## Non-Goals

* Limit regex features to a common subset.

> * Methods of validating regular expressions at the API level. This is theoretically impossible using regex and CEL.

Reviewer 3: ```suggestion
* Define methods of validating regular expressions at the API level. This is theoretically impossible using regex and CEL.
```

(Grammar nit.)

Although out of scope for this proposal, some use cases are:

* Path/URL matching
* URL rewriting
* Header matching

The goal here is to define a common regex syntax and semantics that all implementations should support.
**Individual implementations can support more features.**
**This give us portability guarantees between implementations, but also lets us define conformance tests with regular expressions**

For example, a **hypothetical** path rewrite filter could look something like

```yaml
pathRewrite:
  pattern: ^/foo/(.*)$
  replacement: /bar/\1
```

Since the above pattern and replacement only use features supported by Gateway API Regex, its behavior should be consistent across implementations.

Individual implementations can support more extended regex features, such as RE2:

```yaml
pathRewrite:
  pattern: ^/foo/(?P<name>.*)$
  replacement: /bar/${name}
```

but any portability guarantees are now lost.

In other words, any pattern (and replacement) must implement Gateway API Regex syntax and semantics, but implementations can support additional syntax and semantics as well.

> An unfortunate consequence of this API is that there is no way for users to programatically verify whether their regexes fall under this GEP.

Reviewer 3: Thinking about this some more, I feel like we probably _could_ define some helpers here that would not necessarily guarantee that a regex fits, but could help flag things that don't. For example, a regex that contains any of 

- `(?`
- `${`
- `[[`

not preceded by a `\` is _probably_ a problem. Likewise, `\` followed by a letter that's not an `n` might be a problem. (Worth noting, too, that this list is me reading examples and thinking, _not_ me studying the spec, so take it with a grain of salt.)

I don't think we should say that implementations MUST flag such things in `status`, but it might be worth a MAY or SHOULD...

Author: I just think these rules can get tricky. We are trying to thread the needle between what is not allowed/defined in Gateway API regex, but is allowed in other regexes. And if we have a mechanism to tells people what isn't Gateway API regex (most of the time), they might interpret the converse is true: if there is no problem in the status, the regex is definitely portable. At this point, I'd rather just formally define what Gateway API regex allowed with some implementation of BNF.

Reviewer 3: First off, I should've been explicit that I don't think we should hold this GEP for that work.

But, yes, a BNF definition as a later addition sounds great. 🙂

There is an implicit line between standard and implementation-specific behavior configuration.
That said, this proposal is still an improvement over our current state of `RegularExpression` path matches, which are fully implementation specific and have no conformance tests.

> Further, tools like regex101.com make it easy to check your regex across various implementation.

Reviewer 3: ```suggestion
Further, tools like [regex101.com](https://regex101.com) make it easy to check your regex across various implementation.
```

## Implementation and Support

The most popular Regex engines in the networking world are RE2 and PCRE.

| Proxy          | Engine         |
|----------------|----------------|
| Envoy          | RE2            |
| HAProxy        | PCRE           |
| NGINX          | PCRE           |
| Traefik        | RE2            |

That said, we want to support other languages and engines. For the rest of the document "most modern regex engines" will mean

* Python's `re`
* Ruby's `Regexp`
* Perl (PCRE)
* Go `regexp` (RE2)
* Rust `regex`
* Java `java.util.regex`

Gateway API Regex should be a subset of most modern regex engines.

## Inputs

For Gateway API Regex we only explicitly support strings composed of ASCII characters that are not null, not control characters, and not spaces (other than space and tab) as inputs.

> In other words, only the following code points are supported:

Reviewer 3: Props for correctly using "code points" here. 🙂

* `0x09` (tab)
* `0x20` (space)

> * `0x21-0x7E` (printable characters)

Reviewer 3: One could argue that this should be e.g. `U+0009` but since you've already said that we only support ASCII, I think this representation is fine.

> The reason for this restriction is that most modern regex engines differ in their support for line breaks, unicode, and control characters.

Reviewer 3: ```suggestion
The reason for this restriction is that most modern regex engines differ in their support for line breaks, Unicode, and control characters.
```

Supporting such inputs would add ambiguity and diminish portability with little practical benefits as such characters are rarely used in traffic routing and manipulation 
(notably, [HTTP header values can have `0x80-0xFF` code points](https://datatracker.ietf.org/doc/html/rfc9110#section-5.5), but their use is rare).
For example, in RE2, the `.` character might or might not match line breaks, depending on configuration.
`rust-lang/regex` treats multi-byte unicode characters as a single character, while in RE2, they are treated as individual bytes.
Null and control characters should not show up in traffic routing.

Some examples of valid input strings are

* `foo` (`0x66 0x6F 0x6F`)
* ` foo bar` (`0x20 0x66 0x6F 0x6F 0x20 0x62 0x61 0x72`)
* `foo\tbar` (`0x66 0x6F 0x6F 0x09 0x62 0x61 0x72`)

Some examples of invalid input strings are (implementations need not reject these, but they are not explicitly supported by the spec):

* `foo\nbar` (`0x66 0x6F 0x6F 0x0A 0x62 0x61 0x72`)

> * `ｆ๏๏ ß∆я` (`0xef 0xbd 0x86 0xe0 0xb9 0x8f 0xe0 0xb9 0x8f 0xe2 0x80 0x82 0xc3 0x9f 0xe2 0x88 0x86 0xd1 0x8f`)

Reviewer 3: ```suggestion
* `ｆ๏๏ ß∆я` (`0xef 0xbd 0x86 0xe0 0xb9 0x8f 0xe0 0xb9 0x8f 0xe2 0x80 0x82 0xc3 0x9f 0xe2 0x88 0x86 0xd1 0x8f` using UTF-8)
```

* Backslashes (`\`) retain their special meaning in character classes and can escape special characters, even if those special characters lose their special meaning in character classes. For example, `[\-]` matches `-`, and `[\]]` matches `]`. See [Semantics inside square brackets](#semantics-inside-square-brackets) in the appendix.
* Unescaped square brackets in character classes are undefined (`[[]]`, `[]]`, `[[]`, `[[:alpha:]`, `[[=a=]]`, `[[.ch.]]`).
* Consecutive quantifiers are undefined (`a**`, `a*?`, `a{1,2}?`).
* Any locale-specific behavior will assume the [C/POSIX locale](https://pubs.opengroup.org/onlinepubs/7908799/xbd/locale.html) (e.g. character ordering).

IEEE POSIX ERE is a good common denominator because
* The set of supported features is small (e.g. no backreferences within patterns)
* Broadly compatible across most modern regex engines, especially because we don't have to worry about unicode, line breaks, and control sequences.

Some notable omissions implied from our construction of Gateway API Regex are:
* Backreferences (e.g. `\1`, `\2`, etc.) within patterns.
* Matching flags (e.g. `(?i)` for case-insensitive matching).

> * `\d` for digit, `\w` for word characters, `\s` for whitespace characters, and escaping any character that isn't a special character.

Reviewer 3: It also seems to be the case that `\d`, `\w`, and `\s` are supported across all our common engines? Again, sing out if you found an engine that doesn't support them.

Author: true, but they aren't in POSIX ERE and I figured we'd start out with something minimal.

Reviewer 3: These days, it always annoys me when I have to spell out `[0-9][0-9]*` for stupid implementations of `sed` rather than being able to use `\d`. These are _so_ widely supported that I'd like to just say they're OK. 🙂

That said, this proposal is still an improvement over our current state of `RegularExpression` path matches, which are fully implementation specific and have no conformance tests.
Further, tools like [regex101.com](https://regex101.com) make it easy to check your regex across various implementation.

## Implementation and Support

The most popular Regex engines in the networking world are RE2 and PCRE.

| Proxy          | Engine         |
|----------------|----------------|
| Envoy          | RE2            |
| HAProxy        | PCRE           |
| NGINX          | PCRE           |
| Traefik        | RE2            |

That said, we want to support other languages and engines. For the rest of the document "most modern regex engines" will mean

* Python's `re`
* Ruby's `Regexp`
* Perl (PCRE)
* Go `regexp` (RE2)
* Rust `regex`
* Java `java.util.regex`

Gateway API Regex should be a subset of most modern regex engines.

## Inputs

For Gateway API Regex we only explicitly support strings composed of ASCII characters that are not null, not control characters, and not spaces (other than space and tab) as inputs.
In other words, only the following code points are supported:

> * `0x09` (tab)

Reviewer 4: I understand the need for spaces on headers and URLs matching, but why tabs? maybe this could be more strict as in spaces and printable characters only (which I like very much as initial restriction)

Author: sounds good.

Author: oh again, these are things that we match against, not the actual code points allowed for the pattern itself.

Reviewer 4: ah right, sounds good.

* `0x20` (space)

> * `0x21-0x7E` (printable characters)

Reviewer 4: just some concern from me: http://www.csc.villanova.edu/~tway/resources/ascii-table.html

some printable characters here are slightly concerning from an API field point of view (quotes, etc) so we need to be sure that implementations understand the security implication of this "lack of validation" on this field

Author: @rikatz I think you are misreading this. I'm saying that you can't use Gateway API regex to match crazy unicode/emojis. So if in the future, we have regex header value matching, we're not going to define what happens if your header values contain emojis/control characters.

* whether the first or all matches should be replaced.
* if replacing all matches, repeatedly find the next non-overlapping leftmost-first match, replace it, then continue scanning after the replaced match.
    * This means that some instances of patterns might be skipped if there are overlaps.
    * The resulting string might contain new matches due to replacements.

### Referencing Capturing Groups

In the case that referencing capturing groups is supported, APIs MUST use the `\1`, `\2`, etc. syntax in the replacement string to reference capturing groups.
These can be escaped with an additional backslash (e.g. `\\1` to reference the literal string `\1`).
Note that some regex engines use the `$1`, `$2`, etc. syntax to reference capturing groups, but we choose the `\1`, `\2`, etc. syntax.
Implementations with will need to translate the `\1`, `\2`, etc. syntax to their engine's syntax (e.g. `$1`, `$2`, etc.) if necessary.

Named capturing groups are not explicitly supported, and the syntax for referencing them is undefined.
A capturing group is any expression enclosed in unescaped parentheses.

> The order of capturing groups is determined by the order of the opening parentheses, from left to right in the pattern, starting at 1.

Reviewer 5: I assume supporting `\0` is up to implementations ?

Author: yes, but if its popular enough I don't mind adding it here.

In the third example, the second `aba` still shows up in the output because patterns are not reevaluated after replacements.

Note in the last example, the pattern is `(a|ab)`, the backslash is to escape the `|` in the markdown table.
The output is `cb` rather than `c` because of leftmost-first matching.

## Appendix

### Justification for POSIX ERE exceptions

We use leftmost-first rather than leftmost-longest semantics because
most modern regex engines are leftmost-first and some, such as PCRE, cannot be configured otherwise.

Most modern regex engines retain the special meaning of backslash inside square brackets.

The behavior of unescaped square brackets in character classes varies across implementations.
For POSIX ERE the meaning of unescaped square brackets in characters depends on its position.
For example, in PCRE, the pattern `[][]` means and opening square bracket or a closing square bracket.
However, this pattern is invalid in Rust.

Consecutive usually redundant and may have surprising behaviour.
For example, in the pattern `a+?` is different than `(a+)?` in RE2.

> In the pattern `a+?`, the `?` modifies the `+` quanitifier  be lazy.

Reviewer 4: ```suggestion
In the pattern `a+?`, the `?` modifies the `+` quantifier to be lazy.
```

# GEP-4359: Gateway API Regex

* Issue: [#4359](https://github.com/kubernetes-sigs/gateway-api/issues/4359)

> * Status: Provisional

Reviewer: This should be a Memorandum, not Provisional. We're defining a general thing, not specific fields or use cases, so it doesn't need to go through the Provisional/Implementable/Experimental/Standard flow.

## TLDR

Regular expressions provide a powerful and concise method for traffic routing and manipulation, but present portability challenges due to the variety of regex engines.
The goal of this proposal is to define a common regex syntax and semantics that all implementations should support to ensure portability, but mostly testability.

> We use POSIX ERE (as defined in Chapter 9 of the [The Open Group Base Specifications Issue 8]) as the common denominator, with some feature exclusions.

Reviewer: This link is broken?

* Define specific use cases for regular expressions.
* Define failure modes when a regular expression or input is unsupported. Such validation requires a context-free grammar, and we cannot do that with CEL.

## Introduction/Overview

Regular expressions are a concise mechanism to describe large sets of strings, and for purposes of traffic routing and manipulation.
Unfortunately, the many regular expression engines in the proxy world ([RE2](https://github.com/google/re2), [PCRE](https://www.pcre.org/), and [rust-lang/regex](https://docs.rs/regex/latest/regex/))
present a portability challenge for Gateway API, as different proxies use different engines which support different syntax and semantics.

## Purpose

Regular expressions allow users to concisely express patterns for traffic routing and manipulation.
Although out of scope for this proposal, some use cases are:

* Path/URL matching
* URL rewriting
* Header matching

The goal here is to define a common regex syntax and semantics that all implementations should support.
**Individual implementations can support more features.**

> **This give us portability guarantees between implementations, but also lets us define conformance tests with regular expressions**

Reviewer: This isn't really correct any more. Maybe something more like this?
```suggestion
**This allows us to set a minimum functionality floor that is testable using the conformance suite, and so will be guaranteed to be portable. **
Implementations that support more functionality than the tested conformance MUST inform their users that their regexes may not be portable.
```

## Gateway API Regex Definition

We use IEEE POSIX ERE (as defined in Chapter 9 of the [The Open Group Base Specifications Issue 8]), as the base of Gateway API Regex, with a few exceptions:

* Matches are leftmost-first rather than leftmost-longest (quantifiers are still greedy).
* Backslashes (`\`) retain their special meaning in character classes and can escape special characters, even if those special characters lose their special meaning in character classes. For example, `[\-]` matches `-`, and `[\]]` matches `]`. See [Semantics inside square brackets](#semantics-inside-square-brackets) in the appendix.
* Unescaped square brackets in character classes are undefined (`[[]]`, `[]]`, `[[]`, `[[:alpha:]`, `[[=a=]]`, `[[.ch.]]`).
* Consecutive quantifiers are undefined (`a**`, `a*?`, `a{1,2}?`).
* Any locale-specific behavior will assume the [C/POSIX locale](https://pubs.opengroup.org/onlinepubs/7908799/xbd/locale.html) (e.g. character ordering).
* `\d` matches any digit (`[0-9]`), `\w` matches any word character (`[a-zA-Z0-9_]`), `\W` matches any non-word character `[^\w]`, `\s` any whitespace character (in this context, this is only tabs and spaces).

> IEEE POSIX ERE is a good common denominator because

Reviewer: Need an extra line after this for the website renderer.

* The set of supported features is small (e.g. no backreferences within patterns)
* Broadly compatible across most modern regex engines, especially because we don't have to worry about unicode, line breaks, and control sequences.

> Some notable omissions implied from our construction of Gateway API Regex are:

Reviewer: Need a newline here too.

* Backreferences (e.g. `\1`, `\2`, etc.) within patterns.
* Matching flags (e.g. `(?i)` for case-insensitive matching).
* Hex encoding of characters (e.g. `\xFF` or `\uFFFF`).
* Lazy vs non-lazy matching (e.g. `*` vs `*?`).
* Interval range expressions without starting numbers (`a{,2}`, `a{,}`).

If a regex contains undefined syntax, then the behavior is implementation specific, and implementations are not required to reject such regexes.

### Warnings

Exposing regexes in an API has a lot of sharp corners since its hard to validate the validity of regexes in at the API level.

> **APIs that use Gateway API Regex should provide warnings that regexes outside of this GEP are not portable across implementations and there is no validation surrounding the portability of specific patterns.**

Reviewer: ```suggestion
**APIs that use Gateway API Regex MUST provide warnings that regexes outside of this GEP are not portable across implementations and there is no validation surrounding the portability of specific patterns.**
These warnings MAY be in the API docs, implementation docs, status messages or other places, but they MUST be present.
```

| `^`                                    | _(no match)_ |
| any string with two or more characters | _(no match)_ |

For the equivalent patterns `^[-a]$` and `^[a-]$`:

| Input         | Match        |
| ------------- | ------------ |
| `-`           | `-`          |
| `a`           | `a`          |
| anything else | _(no match)_ |

For the equivalent patterns `^[-.*(){}|^$\\]$` and `^[\-\.\*\(\)\{\}\|\^\$\\]$`:

| Input         | Match        |
| ------------- | ------------ |
| `-`           | `-`          |
| `.`           | `.`          |
| `*`           | `*`          |
| `(`           | `(`          |
| `)`           | `)`          |
| `{`           | `{`          |
| `}`           | `}`          |
| `\|`          | `\|`         |
| `^`           | `^`          |
| `$`           | `$`          |
| `\`           | `\`          |
| anything else | _(no match)_ |

For the pattern `[\-Z]`:

| Input         | Match        |
| ------------- | ------------ |
| `-`           | `-`          |
| `Z`           | `Z`          |
| anything else | _(no match)_ |

## Replacement

Replacement is usually straightforward, but has some semantic ambiguities.

> Given a pattern, a replacement, and an input, we allow higher level APIs to define semantics such as

Reviewer: Newline required after this line.
